File: LeetCode/Medium/0074-search-a-2d-matrix/0074-search-a-2d-matrix.go
func searchMatrix(matrix [][]int, target int) bool {
    //convert to 1d 
    //find mid and find row its in - mid/n
    //find column its in as well - mid % m
    //middle element is matrix[mid/n][mid%n]
    m, n:= len(matrix), len(matrix[0])
    l, r := 0, m*n-1
    for l<=r{
        mid := l+(r-l)/2
        if target == matrix[mid/n][mid%n]{
            return true
        }else if target > matrix[mid/n][mid%n]{
            l = mid + 1
        }else{
            r = mid - 1
        }
    }
    return false
}

File: LeetCode/Medium/0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size(), col = matrix[0].size();
        int l = 0, r = (row * col) - 1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            //only trick here is this to find mid value in matrix
            int midV = matrix[mid / col][mid % col];
            if (target == midV) {
                return true;
            } else if (target > midV) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return false;
    }
};